This PR seeks to add this Hebrew vowel common character file based on the IHBC standard.

yeatersink opened the pull request.

yeatersink commented:
I believe this is what everyone is expecting. @LeonarddeR  @Paulg

paulGeoghegan commented:
Hi, I could be wrong but does this common characters file include the consonants as well as the vowels? If so I don't think this is the correct approach.
If others disagree then I think the files should at least be renamed to he-common-ihbc and then he-common-(other standard).

yeatersink commented:
@paulGeoghegan  I feel like I have failed you. Hebrew vowels are dots that are attached to consonants. Adding to the complexity, hebrew letters have dageshes  which are dots in the consonants. 

If we are going to have a common characters table for the vowels, we have to have the vowels standing alone, and connected to the consonants witht he dageshes, or the table will not produce the correct braille. 

What creates this phenomenon? 

Because of how the Unicode produces the Hebrew, the Unicode could be produced in 1 of 2 ways: 
1 the consonant, the vowel and then dagesh. or 
2 it could be consonant, dagesh, and then the vowel. 

The table with the vowels needs to take these into consideration to ensure the correct braille is produced. 

This problem is why the first HBO table is not currently producing the correct braille. 


I hope this clarifies the problem.

paulGeoghegan commented:
Hi @yeatersink I was just making sure.
In that case then I definitely think re-naming the common character files would be a good idea since one would be he-common and then this new one.
They should both be he-common and then followed by a dash and whatever system they are using.

yeatersink commented:
@LeonarddeR  I made those schanges to the notes and the dot 5.  This table should be good now.

yeatersink commented:
@egli  and @bertfrees  This table is finished but we need to include the hebrew common character table after it is committed to Lib Louis.

yeatersink commented:
@bertfrees  Yes you are correct. The common character file in #1658 is just the consanants. This common character   file has the vowels that is based on the Vowels in the IHBC standard. Not all tables may want to use this vowel system. 


Thus this PR has a list of characters that will be used by the HBO PR that i will be pushing jjust as soon as you approve this one.

bertfrees commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,472 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+# liblouis  comes with ABSOLUTELY NO WARRANTY.
+
+#This is a common character file that supports Hebrew forward  translation. Matthew Yeater, Paul Geoghegan, and Leonard De Ruijter
```
@yeatersink Can you mention "IHBC" and "vowels" in this description? Note that I will remove the metadata section at the top because this an include table, not a top-level table.

bertfrees replied:
This line is indeed causing a compilation error.

bertfrees commented:
I did some cleanup and force-pushed the branch. @yeatersink Can you please check it out? Like before, I moved the tests to hbo.yaml because for the time being it will be the only table in which the new uti files will be included. There are some failing tests. I'm running the tests against two tables (see hbo.yaml):

1. against a table that includes both the consonants and vowels sub-tables:

        include he-common-consonants.uti
        include he-common-vowels-ihbc.uti

2. and I run them against hbo.utb.

For the first table, there is one failing test. For the second table, four tests are failing. This could be because hbo.utb has shortcomings, which will be fixed in a subsequent PR. Or, perhaps I'm wrong in assuming that hbo.utb should also pass these tests?

yeatersink commented:
Yes I can look over it. And yes if you are referencing old HBO table, those test will fail. The new PR will clear up those failing test. Thank you for all your wonderful hope Bert On Nov 26, 2024, at 10:44 PM, Bert Frees ***@***.***> wrote:﻿
I did some cleanup and force-pushed the branch. @yeatersink Can you please check it out? Like before, I moved the tests to hbo.yaml because for the time being it will be the only table in which the new uti files will be included. There are some failing tests. I'm running the tests against two tables (see hbo.yaml):


against a table that includes both the consonants and vowels sub-tables:
 include he-common-consonants.uti
 include he-common-vowels-ihbc.uti



and I run them against hbo.utb.


For the first table, there is one failing test. For the second table, four tests are failing. This could be because hbo.utb has shortcomings, which will be fixed in a subsequent PR. Or, perhaps I'm wrong in assuming that hbo.utb should also pass these tests?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,476 @@
+
+always ִי 35  # hiriq-yod
+always ֵי 3456  # tsere-yod
+always ְּו 346-3  # Shuriq sheva
```
the order of hebrew characters is strange here. Note that Unicode modifiers bind to the character preceding it. A Shuriq is a vav with dagesh (strictly spoken: mapiq). So to make a proper Shuriq, I think both the dagesh and sheva should come after the vav.

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,476 @@
+always ִי 35  # hiriq-yod
+always ֵי 3456  # tsere-yod
+always ְּו 346-3  # Shuriq sheva
+always ּו 346  # shuruq
```
Idem as above
```suggestion
always וּ 346  # shuruq
```

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,476 @@
+always ֵי 3456  # tsere-yod
+always ְּו 346-3  # Shuriq sheva
+always ּו 346  # shuruq
+always ּוְ 346-3  # Shuriq sheva
```
```suggestion
always וְּ 346-3  # Shuriq sheva
```

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,476 @@
+always כְּ 13-3  # KAF with dagesh  sheva
+always תְּ 1256-3  # tav with daggesh sheva
+
+letter Bֳֳ 345  # hataf qamats
```
This line is malformed and will likely break the table

bertfrees replied:
This was fixed.

yeatersink commented:
@LeonarddeR  I made the suggestions on the Shuriq. Note that there are 4 plausible orders of the unicode. I had them in there earlier, but they must have gotten deleted when i moved them out of the other common character file. There are four possibilities because of the wierd way unicode shows up in Hebrew some times, especially in some sources. This way it will give us accurate braille every time.  

@bertfrees  I made the changes on the Metadata you suggested. I hope that I was careful to make sure not to over write your changes. 

Sorry about the commit failures last night. I am only developing software for the past 6 months and still learning how to use Gethub. @paulGeoghegan  had to help me with his impressive Gethub Ninja skills to revert all my stupidity from last night. My ex- wife wishes i could have sat at his feet years ago. Maybe my ignorance would not be on display for the world to see. Thankfully most of us here are blind and understand that we all make mistakes. lol

LeonarddeR commented:
The diff of the pr now contains all sorts of unrelated changes.

bertfrees commented:
@LeonarddeR Working on it. I'll rebase the branch again so that we don't get the unrelated changes anymore.

bertfrees commented:
@LeonarddeR It's all in one new commit now. Please check [f298d34](https://github.com/liblouis/liblouis/pull/1662/commits/f298d34f276797140bbc2693ae48a37886f4245f).

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,480 @@
+always הָּ 5-125-126  # Hey dagesh  qamats
+always וֹ 135  # holam-vav
+always וּ 346  # shuruq
+always ּו 346  # shuruq
```
This line is really wrong and should be removed. The dagesh should not come before the vav.
```suggestion
```

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,480 @@
+always וּ 346  # shuruq
+always ּו 346  # shuruq
+always וְּ 346-3  # Shuriq sheva
+always ְּו 346-3  # Shuriq sheva
```
Same as above.
```suggestion
```

LeonarddeR commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,480 @@
+always ּו 346  # shuruq
+always וְּ 346-3  # Shuriq sheva
+always ְּו 346-3  # Shuriq sheva
+always ּוְ 346-3  # Shuriq sheva
```
Same as above
```suggestion
```

bertfrees commented:
@yeatersink Thanks for the changes. No compilation errors now, but still the failing tests. We'll ignore the failures for hbo.utb for now. The one failure when the tests are run just on he-common-vowels-ihbc.uti is:

> Input:    'בְּרֵאשִׁית'
> Expected: '⠃⠄⠗⠌⠁⠩⠊⠚⠹' (length 9)
> Received: '⠧⠄⠐⠗⠌⠁⠩⠊⠚⠹' (length 10)
> Diff:     Expected '⠃' but received '⠧' in index 0

yeatersink commented:
@l i appreciate your help. I am confused on what you are suggesting?   If you are saying that the line needs to be removed because it is incorrect, then how is it incorrect? 
If you are saying that the  dagesh never appears in the text in that order, then it is clear that you have never read texts where they have. 

Dageshes are utilized to denote emphesis. You may never come across them in your study, but i come across them in my study oftenand they are misplaced in the text from time to time. Especially in rabbinic literature.     

My reason for placing them in that order is to ensure that we have correct braille. 

I posted on the PR in may a link to a slew of articles where there were "unexpected dageshes." 
The only place where dageshes are not found are in final letters.

Am i mis understanding what you are talking about.

LeonarddeR commented:
@yeatersink The issue is quite easy to explain.
Currently, you have the following definitions:
```always ּו 346  # shuruq
letter ב 1236  # bet
```

Now, in an example text, you can find the sequence `בּו`

This should be brailled as 12-2456 (bw), since the dagesh belongs to the bet, not to the vav. The wrong definition will incorrectly assume a shuruq and results in braille 1236-346.

As said above, the dagesh is a modifier character in Unicode, it modifies the consonant before it, not the consonant after it.

If there are cases where dageshes are placed in wrong places in texts, it is up to the text author/publisher to fix their stuff.It is of course fine to take common errors into account to a certain extent, but in this case there is clearly an ambiguity and IMO the table shouldn't support it.

LeonarddeR commented:
May be there is actually someone sighted who can confirm the visual difference between `בּו` and `בוּ`

yeatersink commented:
@LeonarddeR  I agree with you that all the writers/ publishers should fix their problems. I have been screaming that from the roof tops for 15 years now. Every braille translator that has ever tried to work with languages that have multiple Unicode stanzas have done the same.  No publisher that i have spoken to has listened yet. I refer to even Logos bible software. Horrible with accessibility. 

This is exactly why these tables were built in this way.  Your concern is valid though. 

You are indeed correct about the dagesh modifying the letter before it.  That is the way it is supposed to be. 
The question is:
 Is the letter actually pulling the dagesh from the previous letter? You gave a bet with a shuriq as an example.  This is a valid concern. But in all of the tests that we have created, this does not happen.  I have a huge yaml test that pulls from many sources, and the braille is correct. The yaml test will be in the HBO table which I hope will be committed as soon as this pr is committed into the main branch, so the includ common character tables will work.  

To me there are two solutions: 
1 We can remove the lines and have texts from publishers that are broken and lib louis is not appearing to produce the correct braille. 
2 we can leave these in, and allow Lib louis to produce a braille that takes the problems with the texts  from publishers in mine. 

The key question for me is simple: Is Lib Louis producing the correct braille? 

The test that this table failed today was on the Bet with dagesh and shwa.  The table should have produced a 12-3 but produced a 1236-3-5. 
When i looked at the reason for this, it was because i had accidently removed  the bet with the dagesh that was arranged to take problems like this into consideration.

For me brother, I want correct braille, which is why i am working on these projects witht the dilligence that I am. 

Lets allow this to be merged, and then evaluate if this is a problem with the HBO table and tests. If it creates a problem, then it is 5 minutes to remove them. 
The Yaml test in the HBO table will be quite large and will present all types of opportunities to demonstrate if there are problems as you are concerned about. Trust me, I am concerned too.

LeonarddeR commented:
@yeatersink wrote:
> To me there are two solutions: 1 We can remove the lines and have texts from publishers that are broken and lib louis is not appearing to produce the correct braille. 2 we can leave these in, and allow Lib louis to produce a braille that takes the problems with the texts from publishers in mine.

For me, this is either 1. the table is correct and the text is broken and 2 the table is incorrect and the broken text might be correct, but other text might not be. So definitely going for option 1 here.

Curious what@sjlarose and @EricJHarvey have to say here.

EricJHarvey commented:
There are certainly anomalies in typesetting we should try to make display correctly in these tables, but we need to make sure we aren't fixing problems in ways that will deceive us.
The kinds of issues we should fix are those that display properly to sighted readers but don't easily create good braille. 
In regard to the issues concerning waws and dagheshes, placing the daghesh in front of the waw in the unicode will make it attach to the previous consonant. This wouldn't look right visually. 
I can't think of places where this error might crop up in ways that the unicode would place the daghesh first but the intent was to read a shuruq or consonantal doubled waw, but I'm willing to be proven wrong there.
Matt, would you mind reposting that list of unexpected dagheshes so I can look it over?

sjlarose commented:
I am uncomfortable with defining עו as always shuruq, because there are plenty of instances when it is not. We need ways to account for these. This has been a bug in other environments that has caused consternation because it produced shuruqs where vowels should not be present.

 

Likewise, have we got ִי converting properly to dots 35? I recall this being a controversial one. I know that there are instances when you don’t want this butif the table is designed to conform to IHBC that would be the expected behavior.

yeatersink commented:
Firstly, thank you all for chiming in.  Also, Happy Thanks giving. 
@sjlarose  don’t worry, we will never make the Eyon Vov that you presented a shuriq. The vav will never present itself as a shuriq unless it has a dagesh.   I am sure that was a typo. Lol. I hope you are well. 

@EricJHarvey  there are a couple of things. 
1 Please find the link below which presents weird letters with dageshes. Including Alephs and many others. 
https://www.sefaria.org.il/sheets/219202.1?lang=bi&with=all&lang2=en

2 Just because a dagesh is present, does not mean that it is automatically going to attach to the letter in front of it. You can demonstrate this in MS Word. 
3 A dagesh can stand on its own as well as be in a letter. This is demonstrated in the multitude of examples  mentioned in the link above.
4 There are times where items such as dageshes and accents are typed pre positively, and times where they are written post positively. This is another reason for confusion, and mistakes in writing/ typing. This is also a problem with accents that are written within the context of a hiriq- yod. 
5 We have to keep in mind that the BHS is only one example of the MT. There are 13 different versions of it. The key differences are the accents and dageshes.  We also need to consider Qumran manuscripts. Not to mention the thousands and thousands of pieces of Rabbinic literature.
6 The value of recognizing what the  articles above are saying is that they are authors who are taking hand written print documents, and writing them into Unicode. While doing this, they are preserving how things were written by hand. Rabbis utilized dageshes to underscore emphasis, as well as other things. Thus, they would carry these anomalies over into digital material. Further, while Americans who do not read this literature, these instances  of dageshes would be considered anomalies, but where we read them in academic and liturgical settings, these things are relatively common.  
7 We have to understand that Hebrew is not written like a Roman alphabet. Thus, we cannot just assume that things are going to be written in Hebrew the way we think that they would. Sometimes there are written incorrectly on accident, and sometimes they are written incorrectly intentionally.  
8 this is the problem with 3 and 4 stanza Unicode systems. The entire purpose of this table being recreated was to deal with these issues. 
9This table does try to take all of this into consideration. And, if the dagesh poses a problem in our big test which is coming in the HBO pr, we will change it.  We will have to see how lib louis does the output. As far as my preliminary testing is working, it is working fine without any stealing the dageshes by the pre or post positive consonants. 
I am not trying to argue, I am willing to make changes where they make sense. However, suggestions like this do not appear to be taking the entirety of the situation into consideration.  
The above mentioned are just examples of a fewsenarios. However, we cannot consider Everything that may happen with a text. It seems prudent to try to mitigate the potential issues where we are aware of them, and when we have the chance to do so.

yeatersink commented:
I found an unexpected dot 5 on the yod with shwa. I removed it and now tests seem to be looking ok. NVDA is misreading lines from time to time,

sjlarose commented:
Hi Matt. Thank you for your reply. What I was trying to say is there are significant times when vav with a dagesh is not supposed to be red as a vowel. How are you accounting for these? Because a properly done Hebrew translation system should be able to handle both situations.

yeatersink commented:
@sjlarose  That was the most informed question I have ever had on here. Thank you. Your question also helps to prove my points made earlier and yesterday.  

The dagesh has the ability to stand on its own. Additionally, it has the ability to be included in a letter, and in a letter with a vowel.  

The table that I have been building take all of these into account in the following ways:
1 The table has braille for just the consanants, that stand alone, and have braille for the dagesh which also stands alone. This way, if the consanant and the dagesh are written individually, it will give the letter and the dagesh in braille individually. The beuty with this is that it should produce the correct braille in the correct order. 

2 It has the consanant and dagesh in a single stanza, so it produces braille in the same manner. 
3It has the consanant  vowel and dagesh written as one stanza. This is how they are usually written and expected to be written in a perfect world. Thus, the table accounts for this and produces the correct sequence so that they are brailled together as one stanza.  The value here is that it keeps the braille from being separated, especially as it pertains to the dageshes.
4  The table takes into consideration the other plausible sequences that the stanza may be arranged in. In other words, Hebrew that is written incorrectly. The table sequences these and brailles them as a single stanza.  this way, the correct braille is produced, no matter how the stanza is arranged.  

Usually a vav with a dagesh that is not intended to be a shuriq is written in two stanzas. In other words, as individual letters. Thus, this table takes all of these into consideration.  

I can add just about any rule to the table just as long as there is not a conflict in the braille. 

We need to check how Lib Louie produces these tables. I think that we are in good shape though. If i need to make a change, it it is literally a 3 minute job.  


I hope this helps. Happy thanksgiving.

LeonarddeR commented:
@yeatersink wrote:
> I found an unexpected dot 5 on the yod with shwa. 

It looks like you removed the dot but didn't change the comment, which still says that it contains a dagesh.

LeonarddeR commented:
I'm still convinced that the following lines need removal as they will produce incorrect braille as explained in https://github.com/liblouis/liblouis/pull/1662#issuecomment-2504368452. In short: they will produce incorrect braille when preceded by ב, פ, כ or ת.

```
always ְּו 346-3  # Shuriq sheva
always ּו 346  # shuruq
always ּוְ 346-3  # Shuriq sheva
```

I'm also confused about `always וְּ 346-3  # Shuriq sheva` and `always וְּ 346-3  # Shuriq sheva`, Aren't a Shuriq and a Sheva mutually exclusive?

yeatersink commented:
@LeonarddeR  You are mistaken on a couple of your points.
You are mistaken that the lines that you suggested will produce incorrect braille. You seem to make arguments but are presenting no evidence.  My preliminary testing shows that the correct braille is being produced. I have uploaded a YAML file to demonstrate it. 

you are mistaken that the Shva does not appear next to the Shuriq. Shva is both vocal and silent. They often appear next to a Vav and at times a shuriq. I made this argument on the original HBO table, provided evidence to demonstrate that these do indeed occur, especially in old latergic literature.  I have a couple of jobs, doing a phd, and do not have time to do the same research and provide evidence tfor the same questions that are being asked over and over again. 

You are correct on your note on the dagesh. NVDA is driving me crazy in VS code. I had to revert to a different version of NVDA. 

I will make that change on the note later today.

LeonarddeR commented:
> You seem to make arguments but are presenting no evidence.

I provided my motivation in https://github.com/liblouis/liblouis/pull/1662#issuecomment-2504368452 . Also, @EricJHarvey  restated in https://github.com/liblouis/liblouis/pull/1662#issuecomment-2504945436:

> In regard to the issues concerning waws and dagheshes, placing the daghesh in front of the waw in the unicode will make it attach to the previous consonant. This wouldn't look right visually.

If something doesn't look right visually, we should not try to make liblouis account for that. It is up to the author of the source to fix their source.

It seems you need another proof of my point. Please investigate the name דָּוִד and Decide for yourself what that should look like in Braille. Then, consider there that the dagesh is preceeding the vav. Then consider that you wrote the definition `always ּו 346`.

LeonarddeR commented:
In addition, you may want to argue that דָּוִד in this table with this definition passes the tests just fine. That seems like a correct conclusion to me. I suspect that Liblouis translates `דָּ` first, as it matches an always rule. Then, as `דָּ` was already processed, the always rule for `ּו` is no longer processed and does not apply.
However, how do we know that this is always the case?

To look at it from a different angle again. If you can provide evidence that instances of de volgorde `ּו` look properly like a Shuriq visually, I"m definitely your man. If you however admit that the order `ּו` was added solely to deal with texts that do not have their unicode in order, I remain of the opinion that this line does not belong in this table.

sjlarose commented:
In most text that I have seen, the order of entry is consonant vowel dagesh, so placing the dagesh before a character would create a situation in which it could be interpreted as part of the previous letter. additionally, visually the result is the same whether the entry is consonant vowel dagesh or continent dagesh vowel.

yeatersink commented:
@sjlarose   and @LeonarddeR 

Here is why I believe that you are wrong: 

Hebrew is written in Unicode stanza. A stanza can have up to 4 different pieces of code on it. Thus Hebrew should be written with the four following parts
a 1consonant, 2 vowel, 3 dagesh , 4 accent. 
Again, lets be clear, a stanza can have from 1-4 pieces of code on it. 
In a perfect world, they are written correct. However, they are not always written correctly and the stanzas get rearranged. Thus it could be a 1 dagesh, consonant, 3 vowel, 4 accent

To take this problem into consideration and to ensure that the correct braille is presented in either scenario, the table is written in the following way:
ּוְ,05BC+05D5+05B0,201.0,always,Shuriq sheva,⠬⠄
וְּ,05D5+05BC+05B0,201.0,always,Shuriq sheva,⠬⠄

Take a close look at the lines above. Notice that it has the character. The Unicode follows?

What these are is individual pieces of Unicode which make up a 4 part stanza. This is how it looks: 05D5+05BC+05B0. All of these little pieces of Unicode, make up a single Stanza. This stanza is presented in Print as a Shuriq with a shva. 

Because the stanza is written in several ways in our chart, when ever Lib Louis comes across  this stanza, it will produce the desired braille. 

So what we are doing is building the stanza from the Unicode. Then we build the Table for Lib Louis.

Our Table in Lib Louis provides the braille based off of the stanza. Lib louis produces the desired braille, every time. At least when the table is correct. In fact, Every single time the test with lib louis is failing, it is because my test document is wrong, Lib Louis produces the desired braille.  I usually blame NVDA for this because it the dev versions of NVDA do not always read correctly. I love NVDA! 

 What you are concerned about is a dagesh being in front of the consonant, and Lib louis will think that the dagesh is part of the letter when it is part of the preceding consonant, and Lib Louis will produce incorrect braille. 

However, the above presentation of the stanza is why this does not appear to be something that we should be too worried about because this table takes into consideration nearly every possible way the stanza could be arranged, and gives the correct braille. This is how we get correct braille, even when the publisher makes mistakes either intentionally or iunintentionally. 

I hope you had a nice thanks giving. 




Why? Again, because Lib Louis is going to be giving braille based on the stanza, because that is how the Unicode is written. Lib Louis is not going to take part of the Unicode from Stanza A, and make it a part of stanza B.

LeonarddeR commented:
> Hebrew is written in Unicode stanza. A stanza can have up to 4 different pieces of code on it. Thus Hebrew should be written with the four following parts a 1consonant, 2 vowel, 3 dagesh , 4 accent. Again, lets be clear, a stanza can have from 1-4 pieces of code on it. 

While I've never seen the word stanza used with regard to Unicode, this is correct so far. I'd personally use the word glyph instead.

> In a perfect world, they are written correct. However, they are not always written correctly and the stanzas get rearranged. 

I don't think the term of correctness is right here. Both the orderings bet dagesh sheva and bet sheva dagesh are valid here, and they produce exactly the same glyph. That said, there is one order that is considered normalized according to the Unicode spec. See [this documentation](https://unicode.org/reports/tr15/#Norm_Forms).

> Thus it could be a 1 dagesh, consonant, 3 vowel, 4 accent

Now unfortunately, this is really a point where you are, in my opinion, definitely incorrect. In Unicode, the various Unicode characters (you call them parts) do not have equal weight. In Hebrew, the consonants are the main characters, while the vowels are combining/modifier characters. The main character, in this case the consonant, should always come first. Therefore the order you mention (dagesh, consonant, vowel, accent) is not equivalent to the orderings where the consonant comes first.

I would challenge you to find someone who can tell, using his/her eyes, whether the three sequences below look the same. I asked my wife the same thing, and she confirmed that the second combination looks visually different from the first and the third.

`בֶּ    ּבֶ    בֶ`

yeatersink commented:
@LeonarddeR  Tomato Tamato.  Finally you are understanding the argument and the reasoning for the rearrangement of teh unicode stanzas. 
Again, this is why you are wrong this time. Even @sjlarose  can attest that the Hebrew does get mixed around. And it happens alot.  

I am glad your wife is helping you, my wife who also speaks hebrew looks at it and sees the same thing your wife does.  So both of our sighted wives are correct. 

Here is why you are incorrect:

Now, let me tell you that the Unicode does get mixed around, and look right visually. That is how they make mistakes writing Hebrew unintentionally. 

The reason that your wife and my wife can look at them and see that they different, is because our software will generate a character based upon the Unicode stanza. So it is correctly creating a character in print based upon the unicode.  That proves that the concept of the stanza is working!

The point that you are missing now is that the Unicode stanza can be arranged both correctly and incorrectly. This can show itself visually in print or not. Again, @sjlarose  can attest taht the unicode looks right visually, but the word is not brailled correctly because the unicode is arranged incorrectly.  

The point of having these in the table, is that no matter the arrangement of the unicode stqanza, we get correct braille. 
I hope this helps.

LeonarddeR commented:
I'm overly confused. Are you saying that your wife is seeing the same as mine, namely that dagesh bet segol is looking different visually from bet segol dagesh and bet dagesh segol?
The first is incorrect unicode, the second is correct normalized unicode, the latter is correct unnormalized unicode.

sjlarose commented:
Dagesh bet segol would be different because the dagesh would go to the previous consonant.

 

In Unicode, often people type the dagesh as the last thing in the ordering, e.g.

 

דִבַּרְתִּי

 

This is identical to

 

דִבַּרְתִּי

 

The dageshes are properly on the bet and the tav and should never be pushed in Unicode in front of them. If they are pushed in front in Unicode, they will then visually be on the dalet and the bet.

 

That translation should be made on the braille side.

sjlarose commented:
I also have never heard the word stanza used in reference to Unicode

 

Is it possible that Matt’s wife is seeing ּבֶ as correct because it is isolated? I don’t think it would be correct in a word, e.g.

 

דִּבֶר

 

In that situation there is a conflict regatding which character the dagesh belongs to. It will go to the dalet.

yeatersink commented:
@sjlarose  This table does exactly that. It actually does that better because if the Hebrew is written correctly, this table will produce the correct braille.  If the Hebrew is written incorrectly, this table will produce the correct braille. 

Again, we just need to get the table committed so we can test it on a larg YAML test. 

If there any changes that need to be made, I can do that in a matter of minutes.  We need the table to be committed in order to test the HBO table with it.

EricJHarvey commented:
I maintain that when the visual Hebrew presents correctly,  the braille
should also present correctly. But when the visual Hebrew does not present
correctly, the braille should not be correct either. We don't want our
braille to fool us into thinking there is no problem with the text when the
problem is evident to a sighted reader.
Another way of saying this is that our braille should present to us what is
on the page.
Otherwise, we would not be able to accurately assess student work, for
example, because the translator could fix errors the students were making.

There are multiple orders to type characters that present correct visual
Hebrew, but not all orders lead to correct  visual Hebrew. As Lenard and
Sarah have said, each glyph/stanza is a hierarchy, and all characters
classed as vowels or accents bind to the preceding consonant/top-level
character.
That means if you typed the dagesh before the consonant you meant it to go
inside, it would attach to the preceding consonant. This would present
incorrectly to sighted readers and should present incorrectly to braille
readers.
I tested this out with a few words and in each case a dagesh typed
preceding a consonant attached to the previous letter. For example, the
verb צוה, which often appears in the Piel in Biblical Hebrew with a doubled
waw.
Typed waw-dagesh-qamats:
צִוָּה
Typed dagesh-waw-qamats
צִּוָה

that's why I asked for examples—not because I don't understand dagesh and
shuruq rules but because I can't think of any specific cases where these
specific lines of code would be necessary, but I'm willing to be shown
otherwise.

That said, these three lines are probably going to be superseded by lines
higher up in the code, which LibLouis gives priority. It might be worth
running lou_trace to find out if and when these rules are being invoked.

On Sat, Nov 30, 2024 at 11:04 AM yeatersink ***@***.***>
wrote:

> @sjlarose <https://github.com/sjlarose> This table does exactly that. It
> actually does that better because if the Hebrew is written correctly, this
> table will produce the correct braille. If the Hebrew is written
> incorrectly, this table will produce the correct braille.
>
> Again, we just need to get the table committed so we can test it on a larg
> YAML test.
>
> If there any changes that need to be made, I can do that in a matter of
> minutes. We need the table to be committed in order to test the HBO table
> with it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/liblouis/liblouis/pull/1662#issuecomment-2509142948>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIP34HIZTJA34I2LRUOY4HT2DID3PAVCNFSM6AAAAABSAMET2GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBZGE2DEOJUHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
————————————
Eric J. Harvey, Ph.D.
Postdoctoral Fellow in Digital Humanities
Center for Spatial and Textual Analysis
Stanford University

www.blindscholar.com

LeonarddeR commented:
Well said @EricJHarvey! Nothing to add to that.

In the end, I don't think there's much to add to this discussion now, especially since a new release is very near. I'd really like to see this land since it is a massive improvement over the currently released table/approach and we need to decide. From the discussion above, I hope @egli and @bertfrees wille be able to pick a standpoint.

yeatersink commented:
I appreciate the help, and the criticism. I find value in both. As mentioned in other comments, I am willing to make changes when they are needed. 

The problem is that there is much more than opinion that is at stake here.  You are not taking into consideration the arguments that I made earlier regarding the multitude of texts and areas where the Unicode of a particular stanza may be misarranged.  This happens all the time especially when a document gets converted from one format to another, from csv to html and vice versa. I already provided the links to the sources on this Pr and on the PR in May.  

Indeed, the YAML on the HBO table also produces test from a multitude of sources. The reason for this is to ensure that we are receiving correct braille. As I mentioned several times, there are times when the Unicode is misarranged accidently, and intentionally. I want correct braille. The way this table is set up is to ensure that we get correct braille.

The big argument now is that you want incorrect braille. Well, the entire reason for these tables is to stop incorrect braille. I am tired of incorrect braille. This table seeks to produce correct braille.  

The other thing that is the most disappointing is the arguments in May about how the “rest of the world uses Hebrew, not just you.” This was a great statement because it causes to think how others who are using this table will think or respond to the places where the braille is incorrect. Do you think that they are going to stop and think that the Hebrew is written incorrectly even though it looks correct visually?
No, they won’t. They will look at it appearing correct in print, but will assume that Lib Louis is producing incorrect braille. 

I would like to remind everyone that the reason  for re-working the Hebrew tables was to deal with he problem of the misarrangement of the Unicode.  Leonard make a common character file but then closed it. We had no choice but to take up the work because he closed the Pr. The suggestions have been helpful, and the suggestions have demonstrated that we have worked together to produce a nice table. 

In my opinion, I think that we have crossed over to something this was not intended to be when we have gotton to the place where we ask the folks at Lib Louis to pick a side, and choose correct braill or incorrect braille. 

I need this for my work. I have spent hours and hours working to produce correct braille. For the fifth time, Can we please move forward and actually see how Lib Louis is going to handle this table with the HBO YAML file. For the fifth time, if there any changes that need to be made because the table is not performing as expected, I will make the changes.

yeatersink commented:
@LeonarddeR  @EricJHarvey  @sjlarose  I removed the characters where the unicode is not arranged correctly. 
This Table should now contain only the Characters with the correct Unicode. 

After thinking about it it, I realized that it was extremely important to make these changes.

Your position is very incorrect. Your arguments do not engage evidence I provided several times. 

The fact that you make the arguements that you do demonstrate that you do not engage with the vastness of Hebrew literature. Nor do you understand the critical issues I am raising. 

The fact is that I have been living with incorrect braille for the past 15 years, I can live with it for 3 more months. 

Your comments demonstrate that you need this table more than I do. 

It is probably more important to make this tablenow.  and when I need to update the table later, I can easily do that.

EricJHarvey commented:
Can't check this right now, but just to clarify: did you remove all lines with incorrect unicode order or just the three specific lines with the dagesh before the waw? Those are the only lines I was referencing above. As Lenard also said, this table is an improvement over what came before, and I don't think those three lines are worth holding it up over. They are rules that will be invoked rarely, and we can see in practice if they cause problems one way or the other.The deeper issue regarding whether the table should produce correct (meaning orthographically and grammatically correct) braille or accurate braille (meaning what is on the page or screen) is something we will need to keep discussing and something those with expertise on other tables can help with. Eric J. Harvey, Ph.D.He/hisAffiliated ResearcherCenter for Spatial and Textual AnalysisStanford Universitywww.blindscholar.comOn Dec 1, 2024, at 8:31 AM, yeatersink ***@***.***> wrote:﻿
@LeonarddeR  @EricJHarvey  @sjlarose  I removed the characters where the unicode is not arranged correctly.
This Table should now contain only the Characters with the correct Unicode.
After thinking about it it, I realized that it was extremely important to make these changes.
Your position is very incorrect. Your arguments do not engage evidence I provided several times.
The fact that you make the arguements that you do demonstrate that you do not engage with the vastness of Hebrew literature. Nor do you understand the critical issues I am raising.
The fact is that I have been living with incorrect braille for the past 15 years, I can live with it for 3 more months.
Your comments demonstrate that you need this table more than I do.
It is probably more important to make this tablenow.  and when I need to update the table later, I can easily do that.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

yeatersink commented:
@EricJHarvey  This is yet another time where you offer critisism without evaluating the content.  Again you flip on your position with out evaluating the table or the evidence for the content of the table. . This is extremely unprofessional in my opinion. 

You commented about expertise. I have 3 masters degrees, been in a Michon learning Rabinic Literature for 6 years, and currently working on a Phd where we are developing ground breaking accessibility in Ancient languages. Expertise in the language is not the problem on my end.  

The problem is that you and your team @sjlarose  and @LeonarddeR  have continually refused to work on this project together after I have consistently called and sent emails requesting colaberation.  You wont work on the table in any constructive manner, But are willing to criticize with out engageing evidence, and introduce massive changes to this table With in literally hours from the release date of lib Louis. 

The problem is that even after being presented evidence by articles and websites, you and your team still make arguements, that clearly demonstrate that you are not engageing with the material. 

The problem is that even when you are proven wrong by the evidence, your final argument is that you would be happy with incorrect braille. 

I make mistakes, and I need a certain amount of grace when I am making mistakes and learning from them. For that reason, I am the first to offer understanding. But after the past few days, and after the requests to make Massive and unnecessary changes to this table so close to a release date, it appears that something is not right here. 

The blatent refusal to look and intelegently evaluate the evidence for the decisions on the table, and arguing that incorrect braille is prefered, I am becoming more and more convinced that this is a mission      to try to sabotage the success of this table and project. 

This behavior is not indicative of a open-source community. It is a shame that @LeonarddeR  would ask the folks at Lib Louis to choose a side, and that be your side, even if it means incorrect braille. 

@bert and @egli  I am ready and at your disposal if there are any changes that need to be made to this table. or the HBO table. I need this for my work, and I am asking you to please move forward with this pr.

yeatersink commented:
@egli  @bertfrees  there was many back translation conflicts in the previous table, so I removed them. For example, dots 12-3 could have backtranslated in 1 or 2 ways. 1 bet, dagesh, shva or 2 bet, shva, dagesh. This would have created more problems.  So i removed them, and will work that out in a future table update. 

I could have just put a noback opcode, but for the sake of moving forward with this table in this release, I figured that it might be better to just remove the problematic characters. Unless you want them back in, I can do what ever you want.  

As it stands, this table will give any one using it a fully functioning forward and back translation. As long as their text complies with normal Unicode as it pertains to Hebrew. 

Thank you for all your help.

EricJHarvey commented:
Matt,I think you misunderstood me. I was not asking you to do anything in my last message. I was only asking for clarification. I was confused by what you meant by your message and I was atually worried you were going overboard and doing too much right before the release. No one is trying to sabotage this table. We all want good, working Hebrew tables.Just to be clear, my position right now is to proceed with this release and we'll address necessary changes on the other side.Eric J. Harvey, Ph.D.He/hisAffiliated ResearcherCenter for Spatial and Textual AnalysisStanford Universitywww.blindscholar.comOn Dec 1, 2024, at 1:42 PM, yeatersink ***@***.***> wrote:﻿
@egli  @bertfrees  there was many back translation conflicts in the previous table, so I removed them. For example, dots 12-3 could have backtranslated in 1 or 2 ways. 1 bet, dagesh, shva or 2 bet, shva, dagesh. This would have created more problems.  So i removed them, and will work that out in a future table update.
I could have just put a noback opcode, but for the sake of moving forward with this table in this release, I figured that it might be better to just remove the problematic characters. Unless you want them back in, I can do what ever you want.
As it stands, this table will give any one using it a fully functioning forward and back translation. As long as their text complies with normal Unicode as it pertains to Hebrew.
Thank you for all your help.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

LeonarddeR commented:
I don't feel the need to respond to all the insults posted in https://github.com/liblouis/liblouis/pull/1662#issuecomment-2510256280 .

In short.

1. The last intervention in last commit c198288020208b7a75d4eee0532 is too drastic in my opinion. I could make arguments for this but they would undoubtedly be ignored or labeled as untrue. Note that there is a difference between incorrect unicode that also produces incorrect glyphs, or just unicode that produces the proper glyph but is just not normalized. The latter is *not* incorrect per se.

2. The last commit introduces a massive amount of errors. Checking a commit before pushing it is, in my opinion, something that is a virtue within a professional attitude, unfortunately it is not the first time that I have had to point this out. But hey, let's not talk about professionalism, because opinions also differ on this. A short overview about what's wrong:
    ```
    always אְ 5-1-3  # alef with  sheva
    always בְ 12-3  # bet with sheva
    always ג򰖸ּ 5-1245-126  # Gimel dagesh  qamats
    always הְ 5-125-3  # hey with sheva
    ```

yeatersink commented:
@LeonarddeR  I believe those typos are fixed. My mistake.  Can you please look over this real quick and check to be sure.

yeatersink commented:
@LeonarddeR  I cleaned up these notes. But to revert the changes would create more problems than it is worth. This table will work and produce forward adn back translation. 

I can go back and add in the characters I took out, but that is going to take a few minutes. If this has the chance to be committed and released today, I would rather have this released.

LeonarddeR commented:
you didn't fix all the notes, for example you removed a line in the file header.
In addition, you now go from one extreme to the other. On Saturday we had the situation that some rules in the table clearly contained incorrect unicode (gadesh vav). Now you have made the choice to limit the table to normalized unicode. For example, there is now only bet segol dagesh, while bet dagesh segol has been removed. Again, ***this is valid unicode***, it is just not normalized.
I can live with this table only containing normalized unicode, but that requires normalization to be done by the screen reader. NVDA has this functionality, so we should be able to use this table just fine with NVDA. With other screen readers, removing this line will produce incorrect braille. Again, I would rather see us go back to the situation before Sunday.

LeonarddeR commented:
1. There is still the line `# according to the IHBC. This table supports forward translation.` in the header that is missing the previous line.
2. There is no need to have a definition of consonant with only sheva, they can be removed. E.g. `always אְ 1-3  # alef with  sheva`

yeatersink commented:
@LeonarddeR  I fixed the line in metadata. As far as the consonant with shva, I think we should leave it. Just because we are forcing everything else, when Lib Louis sees these as one single stanza, it will give the correct braille every time. 

If it creates problems, I will remove iit in the next realease, but I think that this will produce what we are looking for. 
Especially since we are trying to do back translation.

LeonarddeR commented:
As for the consonants with sheva, they probably won't sit in the way.
As we are approaching a deadline, I think we can leave this as is.
However, I would like to note again that I continue to have difficulty with the first two noback always instances, notably:
```
noback always ּו 346  # shuruq
noback always ּוְ 346-3  # Shuriq sheva
```
The unicode order is evidently incorrect. I will therefore not hesitate to request the removal of these rules in a follow-up pull request. But let's not start the discussion over again and then make this discussion block this entire pr.

yeatersink commented:
@LeonarddeR  I removed those lines. This table should be good to go.

LeonarddeR reviewed (approved):
Great, thanks a lot!
    @bertfrees and @egli, this should be ready to go now, apart from the unrelated changes in the diff. I guess a rebase is in order.

sjlarose commented:
I agree that the consonants with sheva are not needed. The only time these are needed is in the case of dagesh. Otherwise sheva a separate character that Unicode will treat so.

bertfrees commented:
You have to excuse me for not reading all the comments. The multitude of emails was a bit too overwhelming for me last night so I procrastinated. I'm glad that you were able to take into account all the feedback in time, and you are all comfortable releasing this table in the current state, and adding further improvement after evaluating it.

I have cleaned up the branch so it doesn't have 1000 commits anymore and no unrelated changes. This should make it possible to review.

The only thing keeping me from pressing the "merge" button are the failing tests. Also the expected braille looks wrong to me because it contains print characters. These are the failures: 

    Input:    'בְּרֵאשִׁית'
    Expected: '⠃⠄ר⠌א⠩⠊ית' (length 9)
    Received: '⠃⠄⠗⠌⠁⠩⠊⠚⠹' (length 9)
    Diff:     Expected 'ר' but received '⠗' in index 2

    Input:    'אֱלֹהִים'
    Expected: 'א⠢ל⠕ה⠔ם' (length 7)
    Received: '⠁⠢⠇⠕⠓⠔⠍' (length 7)
    Diff:     Expected 'א' but received '⠁' in index 0

    Input:    'הָאָרֶץ'
    Expected: 'ה⠣א⠣ר⠑ץ' (length 7)
    Received: '⠓⠣⠁⠣⠗⠑⠮' (length 7)
    Diff:     Expected 'ה' but received '⠓' in index 0

    Input:    'הָיְתָה'
    Expected: 'ה⠣⠚⠄ת⠣ה' (length 7)
    Received: '⠓⠣⠚⠄⠹⠣⠓' (length 7)
    Diff:     Expected 'ה' but received '⠓' in index 0

    Input:    'וְהָאָרֶץ'
    Expected: '⠺⠄ה⠣א⠣ר⠑ץ' (length 9)
    Received: '⠺⠄⠓⠣⠁⠣⠗⠑⠮' (length 9)
    Diff:     Expected 'ה' but received '⠓' in index 2

    Input:    'הַמָּיִם'
    Expected: 'ה⠉⠐⠍⠣⠚⠊ם' (length 8)
    Received: '⠓⠉⠐⠍⠣⠚⠊⠍' (length 8)
    Diff:     Expected 'ה' but received '⠓' in index 0

Firstly, note that the tests are in the file hbo.yaml, not anymore in the separate YAML file. I did this because for now the only place where the new file he-common-vowels-ihbc.uti will be used is in the Classical Hebrew table. When this changes later we can reorganise the tests.

Also note that he-common-vowels-ihbc.uti currently includes he-common-consonants.uti. Perhaps the reason for having the print characters in the expected braille, was because these characters are the consonents? Well, in that case, because the consonants are also covered by the vowels table, it is not needed to do that. It's better to have all braille characters in the expected result.

@yeatersink If you can fix this issue, I think  we're ready to merge.

Since it says in the table header that the table supports forward and back translation, it would be good to also prove that in the tests. But given the little time we have left, that's something we can do in the next release.

yeatersink commented:
@bertfrees  Working on it now.

yeatersink commented:
@bertfrees  I think that fixed it?

yeatersink commented:
@LeonarddeR  I made those changes on the Nun and Tav with shva.

yeatersink commented:
@bertfrees  @LeonarddeR  I believe that everything on this one is fixed.

bertfrees commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,497 @@
+always מֶּ 5-134-15  # mem  with dagesh and segol
+always מַּ 5-134-14  # mem  with dagesh and patah
+always מָּ 5-134-126  # mem dagesh  qamats
+always נְ 1345-3  # nun with  dagesh sheva
```
Done

bertfrees commented on tables/he-common-vowels-ihbc.uti:
```diff
@@ -0,0 +1,497 @@
+always שָׂ 156-126  # sin with qamats
+always שָּׂ 5-156-126  # sin  qamats
+always שְׁ 146-3  #  Shin with dot and sheva
+always תְ 1456-3  # tav with dagesh sheva
```
Done
